/*
 * Copyright (c) 2025-2026, @aquamarine5 (@海蓝色的咕咕鸽). All Rights Reserved.
 * Author: aquamarine5@163.com (Github: https://github.com/aquamarine5) and Brainspark (previously RenegadeCreation)
 * Repository: https://github.com/aquamarine5/ChaoxingSignFaker
 */

package org.aquamarine5.brainspark.chaoxingsignfaker.screen

import android.content.Intent
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.hapticfeedback.HapticFeedbackType
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalFocusManager
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.net.toUri
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import kotlinx.serialization.Serializable
import org.aquamarine5.brainspark.chaoxingsignfaker.BuildConfig
import org.aquamarine5.brainspark.chaoxingsignfaker.R
import org.aquamarine5.brainspark.chaoxingsignfaker.api.ChaoxingHttpClient
import org.aquamarine5.brainspark.chaoxingsignfaker.components.CustomizeClientCard
import org.aquamarine5.brainspark.chaoxingsignfaker.ui.theme.FontGilroy
import org.aquamarine5.brainspark.chaoxingsignfaker.utilities.LocalSnackbarHostState
import org.aquamarine5.brainspark.chaoxingsignfaker.components.AppUpdateCard
import org.aquamarine5.brainspark.chaoxingsignfaker.utilities.chaoxingDataStore
import org.aquamarine5.brainspark.chaoxingsignfaker.utilities.displaySnackbar
import org.aquamarine5.brainspark.chaoxingsignfaker.utilities.reportLocalError
import org.aquamarine5.brainspark.chaoxingsignfaker.utilities.snackbarReport

@Serializable
data class LoginDestination(
    val isFailureNetworkRedirect: Boolean = false
)

@Composable
fun LoginPage(
    destination: LoginDestination,
    navToCourseListDestination: () -> Unit
) {
    var phoneNumber by remember { mutableStateOf("") }
    var password by remember { mutableStateOf("") }
    val coroutineScope = rememberCoroutineScope()
    val context = LocalContext.current
    val hapticFeedback = LocalHapticFeedback.current
    val snackbarHost = LocalSnackbarHostState.current
    val focusManager = LocalFocusManager.current
    Column(
        modifier = Modifier
            .padding(16.dp, 16.dp, 16.dp, 0.dp)
            .fillMaxWidth(),
        verticalArrangement = Arrangement.Center,
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Text(
            buildAnnotatedString {
                append("随地大小签（")
                withStyle(
                    SpanStyle(
                        fontFamily = FontGilroy,
                        fontSize = 16.sp
                    )
                ) {
                    append("ChaoxingSignFaker")
                }
                append("）需要你的学习通账号信息\n请登录你的学习通账号")
            },
            fontWeight = FontWeight.Bold,
            fontSize = 18.sp,
            textAlign = TextAlign.Center,
            modifier = Modifier.align(Alignment.CenterHorizontally)
        )
        Spacer(modifier = Modifier.height(20.dp))
        Text("输入手机号：")
        OutlinedTextField(
            value = phoneNumber,
            onValueChange = {
                phoneNumber = it
            },
            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
            modifier = Modifier.fillMaxWidth()
        )
        Spacer(modifier = Modifier.height(8.dp))
        Text("输入密码：")
        var isPasswordVisible by remember { mutableStateOf(false) }
        OutlinedTextField(
            value = password,
            onValueChange = { password = it },
            visualTransformation = if (isPasswordVisible) {
                VisualTransformation.None
            } else {
                PasswordVisualTransformation()
            },
            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Password),
            modifier = Modifier.fillMaxWidth(),
            trailingIcon = {
                IconButton(onClick = {
                    isPasswordVisible = !isPasswordVisible
                }) {
                    Icon(
                        if (isPasswordVisible) painterResource(R.drawable.ic_eye) else painterResource(
                            R.drawable.ic_eye_closed
                        ), null
                    )
                }
            }
        )
        Spacer(modifier = Modifier.height(16.dp))
        Button(
            onClick = {
                focusManager.clearFocus()
                coroutineScope.launch {
                    runCatching {
                        ChaoxingHttpClient.create(phoneNumber, password, context)

                    }.onFailure {
                        it.snackbarReport(
                            snackbarHost,
                            coroutineScope,
                            "登录失败",
                            hapticFeedback
                        )
                    }.onSuccess {
                        if (ChaoxingHttpClient.instance != null) {
                            snackbarHost.displaySnackbar("登录成功", coroutineScope)
                            navToCourseListDestination()
                        }
                    }
                }
            },
            modifier = Modifier.fillMaxWidth()
        ) {
            Text("登录")
        }
        Spacer(modifier = Modifier.height(8.dp))

        if (destination.isFailureNetworkRedirect) {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .verticalScroll(rememberScrollState())
                    .padding(0.dp, 5.dp, 0.dp, 8.dp)
                    .border(
                        BorderStroke(2.dp, MaterialTheme.colorScheme.onErrorContainer),
                        shape = RoundedCornerShape(8.dp)
                    ),
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                Card(
                    shape = RoundedCornerShape(16.dp),
                    colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.errorContainer)
                ) {
                    fun retryAutoLogin() {
                        coroutineScope.launch {
                            runCatching {
                                ChaoxingHttpClient.loadFromDataStore(
                                    context.chaoxingDataStore.data.first(),
                                    context
                                )
                            }.onSuccess {
                                if (ChaoxingHttpClient.instance != null) {
                                    snackbarHost.displaySnackbar("登录成功", coroutineScope)
                                    navToCourseListDestination()
                                }
                            }.onFailure {
                                it.snackbarReport(
                                    snackbarHost,
                                    coroutineScope,
                                    "登录失败",
                                    hapticFeedback
                                )
                            }
                        }
                    }
                    Column(modifier = Modifier.padding(6.dp)) {
                        Text("自动登录失败，可能是遇到了网络问题或修改了学习通密码。")
                        Button(
                            onClick = {
                                hapticFeedback.performHapticFeedback(HapticFeedbackType.ContextClick)
                                retryAutoLogin()
                            }, modifier = Modifier
                                .fillMaxWidth()
                                .padding(4.dp)
                        ) {
                            Text("尝试重新自动登录")
                        }
                        CustomizeClientCard {
                            retryAutoLogin()
                        }
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            Icon(painterResource(R.drawable.ic_info), null)
                            Text(
                                "如果登录账号持续出现问题，请尝试更新应用版本。",
                                modifier = Modifier
                                    .padding(4.dp, 0.dp)
                                    .fillMaxWidth()
                            )
                        }
                        Spacer(modifier = Modifier.height(4.dp))
                        AppUpdateCard()

                        Spacer(modifier = Modifier.height(4.dp))
                        Button(
                            onClick = {
                                runCatching {
                                    hapticFeedback.performHapticFeedback(HapticFeedbackType.ContextClick)
                                    context.startActivity(Intent(Intent.ACTION_SEND).apply {
                                        setData("mailto:aquamarine5forever@gmail.com".toUri())
                                        putExtra(Intent.EXTRA_EMAIL, "aquamarine5forever@gmail.com")
                                        putExtra(Intent.EXTRA_CC, "aquamarine5forever@gmail.com")
                                        putExtra(
                                            Intent.EXTRA_SUBJECT,
                                            "Send to ChaoxingSignFaker:\n"
                                        )
                                        putExtra(Intent.EXTRA_TEXT, "Your content:")
                                    })
                                }.onFailure {
                                    snackbarHost.displaySnackbar("无法打开邮件应用", coroutineScope)
                                    it.reportLocalError()
                                }
                            },
                            shape = RoundedCornerShape(18.dp),
                            modifier = Modifier.fillMaxWidth(),
                            colors = ButtonDefaults.buttonColors(containerColor = Color(0xFFC08EAF))
                        ) {
                            Row(
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .padding(3.dp),
                                verticalAlignment = Alignment.CenterVertically
                            ) {
                                Icon(
                                    painterResource(R.drawable.ic_mail),
                                    contentDescription = "mail"
                                )
                                Spacer(modifier = Modifier.width(8.dp))
                                Text(buildAnnotatedString {
                                    append("还是有问题？\n联系作者发送邮件到：")
                                    withStyle(
                                        SpanStyle(
                                            fontFamily = FontGilroy,
                                            fontSize = 14.sp
                                        )
                                    ) {
                                        append("aquamarine5forever")
                                    }
                                    withStyle(SpanStyle(fontWeight = FontWeight.Bold)) {
                                        append("@")
                                    }
                                    withStyle(
                                        SpanStyle(
                                            fontFamily = FontGilroy,
                                            fontSize = 14.sp
                                        )
                                    ) {
                                        append("gmail.com")
                                    }
                                })
                            }
                        }
                        Spacer(modifier = Modifier.height(4.dp))
                        Text(
                            "ChaoxingSignFaker versionName:${BuildConfig.VERSION_NAME}, versionCode: ${BuildConfig.VERSION_CODE}, buildDate: ${BuildConfig.releaseDate}",
                            fontSize = 10.sp,
                            lineHeight = 12.sp,
                            color = Color.Gray
                        )
                    }
                }
            }
        }
    }
}